src/NumericalAlgorithms/Interpolation/LinearInterpolationND.hpp

```diff
@@ -0,0 +1,374 @@
+                ...)> = nullptr>
+  inline std::array<double, sizeof...(whichVariables)> interpolate(
+      Xt const&... xin) const {
+    static_assert(
+        sizeof...(Xt) == Dimension,
+        "You need to provide the correct number of interpolation points");
+    static_assert(sizeof...(whichVariables) <= NumberOfVariables,
+                  "You are trying to interpolate more variables than this "
+                  "container holds.");
+
+    int dimension = -1;
+    auto const indexArray = std::array<size_t, Dimension>{
+        (++dimension, findIndex(dimension, xin))...};
+
+    return std::array<double, sizeof...(whichVariables)>{
+        (RecursiveInterpolationKernel<
+            decltype(*this), Dimension - 1, whichVariables,
+            CompressedIndexRecursionInitial_t, Xt...>::interpolate(*this,
+                                                                   indexArray,
+                                                                   xin...))...};
+  };
+
+  template <
+      size_t N, typename... Xt,
+      Requires<(std::is_convertible_v<typename std::remove_cv_t<Xt>, double> &&
+                ...)> = nullptr>
+  inline std::array<double, N> interpolate(std::array<size_t, N> whichVariables,
+                                           Xt const&... xin) const {
+    static_assert(
+        sizeof...(Xt) == Dimension,
+        "You need to provide the correct number of interpolation points");
+    static_assert(N <= NumberOfVariables,
+                  "You are trying to interpolate more variables than this "
+                  "container holds.");
+
+    int dimension = -1;
+    auto const indexArray = std::array<size_t, Dimension>{
+        (++dimension, findIndex(dimension, xin))...};
+
+    std::array<double, N> interpolatedValues{};
+
+    for (int i = 0; i < N; ++i)
+      interpolatedValues[i] = RecursiveInterpolationKernelGeneral<
+          decltype(*this), Dimension - 1, CompressedIndexRecursionInitial_t,
+          Xt...>::interpolate(*this, indexArray, whichVariables[i], xin...);
+
+    return interpolatedValues;
+  };
+
+  template <size_t N, size_t... I>
+  inline std::array<double, N> interpolate_helper(
+      std::array<size_t, N> whichVariables, std::array<double, Dimension>& xin,
+      std::index_sequence<I...>) const {
+    return interpolate(whichVariables, xin[I]...);
+  };
+
+  template <size_t N, typename Indices = std::make_index_sequence<Dimension>>
+  inline std::array<double, N> interpolate(
+      std::array<size_t, N> whichVariables,
+      std::array<double, Dimension>& xin) const {
+    return interpolate_helper(whichVariables, xin, Indices{});
+  };
+
+  LinearSpanInterpolationND() = delete;
+
+  LinearSpanInterpolationND(gsl::span<double const> __y,
+                            std::array<gsl::span<double const>, Dimension> __x,
+                            std::array<size_t, Dimension> __numberOfPoints)
```

Comment:
we typically use the `Index<Dim>` class for representing number of points/extents. This has a `product()` member function that calculates the total number of points, allows slicing, etc.